File: zotprep/review.py
"""Interactive confirm loop for anything the accept gate would not pass.

This is the last mile to 100%. The gate is deliberately strict, so what lands
here is genuinely ambiguous — a handful of references at most. Each decision is
written to the corrections table, so the same reference never asks twice, in this
manuscript or any future one.
"""
from __future__ import annotations

from .cache import Cache
from .models import Candidate, ParsedRef, Resolution
from .resolver import item_from_text as book_from_text


def _describe(c: Candidate) -> str:
    who = c.corporate or (", ".join(a.title() for a in c.authors[:3]) or "?")
    ident = c.doi or (f"PMID:{c.pmid}" if c.pmid else "no identifier")
    loc = f"v{c.volume or '?'}({c.issue or '?'}):{c.first_page or '?'}"
    return (f"{c.title[:88]}\n        {who} | {c.journal or c.publisher or '?'} | "
            f"{c.year or '?'} | {loc} | {ident} | {c.item_type} | {sorted(c.providers)}")


def review_loop(
    refs: dict[int, ParsedRef], results: dict[int, Resolution], cache: Cache
) -> None:
    pending = [n for n in sorted(results) if results[n].status not in ("ACCEPTED",)]
    if not pending:
        print("Nothing to review — every reference passed the accept gate.")
        return

    print(f"\n{len(pending)} reference(s) need a decision.\n")
    for n in pending:
        res, ref = results[n], refs[n]
        print("=" * 78)
        print(f"[{n}] {ref.raw}")
        if res.reason:
            print(f"  why: {res.reason}")

        options: list[Candidate] = []
        if res.candidate:
            options.append(res.candidate)
        options += [c for c in res.alternatives if c not in options][: 5 - len(options)]

        for i, c in enumerate(options, 1):
            marker = " (current pick)" if c is res.candidate else ""
            print(f"  {i}){marker} {_describe(c)}")
        if not options:
            print("  (no candidates found)")

        prompt = "  choose 1-%d / [s]kip / [d]oi / " % len(options) if options else "  [s]kip / [d]oi / "
        if ref.is_book:
            prompt += "[b]uild from reference text / "
        prompt = prompt.rstrip("/ ") + ": "

        while True:
            try:
                ans = input(prompt).strip().lower()
            except EOFError:
                print("\n  (no input available — skipping)")
                ans = "s"
            if ans in ("s", ""):
                print("  skipped — stays flagged in the document")
                break
            if ans == "b" and ref.is_book:
                cand = book_from_text(ref)
                res.candidate, res.status, res.tier, res.confidence = cand, "FROM_TEXT", "book-from-text", 0.80
                cache.put_correction(ref, cand)
                print("  accepted from reference text")
                break
            if ans == "d":
                doi = input("    DOI: ").strip()
                if doi:
                    cand = Candidate(provider="manual", title=ref.title, doi=doi,
                                     authors=list(ref.authors), corporate=ref.corporate,
                                     year=ref.year, journal=ref.journal, volume=ref.volume,
                                     issue=ref.issue, first_page=ref.first_page,
                                     last_page=ref.last_page,
                                     item_type="book" if ref.is_book else "journalArticle")
                    res.candidate, res.status, res.tier, res.confidence = cand, "ACCEPTED", "manual", 1.0
                    cache.put_correction(ref, cand)
                    print(f"  accepted {doi}")
                    break
            if ans.isdigit() and 1 <= int(ans) <= len(options):
                cand = options[int(ans) - 1]
                res.candidate, res.status, res.tier, res.confidence = cand, "ACCEPTED", "manual", 1.0
                cache.put_correction(ref, cand)
                print(f"  accepted {cand.doi or cand.title[:50]}")
                break
            print("  ?")
    print("=" * 78)

File: web/src/review.js
/**
 * Port of zotprep/review.py — the confirm step for anything the accept gate
 * would not pass.
 *
 * The gate is deliberately strict, so what lands here is genuinely ambiguous —
 * a handful of references at most. The CLI asks at a prompt; the browser asks
 * with a panel of options. The decisions themselves, and what each one does to
 * the Resolution, are identical:
 *
 *   pick a candidate      -> ACCEPTED, tier "manual",         confidence 1.0
 *   paste a DOI           -> ACCEPTED, tier "manual",         confidence 1.0
 *   build from the text   -> FROM_TEXT, tier "book-from-text", confidence 0.80
 *   skip                  -> unchanged; stays flagged in the document
 *
 * One deliberate difference: the CLI reviews everything that is not ACCEPTED,
 * which includes FROM_TEXT entries. Here only REVIEW entries are offered, so a
 * run asks about the references that resolved to nothing and nothing else.
 *
 * Decisions live in memory for the session. The CLI writes each one to its
 * corrections table so the same reference never asks twice; this build does not
 * persist anything but credentials, so closing the tab forgets them.
 */
import { Candidate } from './models.js';
import { itemFromText } from './resolver.js';

/** References this build will ask about. */
export function pending(results) {
  return [...results.keys()].sort((a, b) => a - b)
    .filter((n) => results.get(n).status === 'REVIEW');
}

/**
 * The candidates to offer, matching review.py: the current pick first if there
 * is one, then alternatives, five in total.
 */
export function optionsFor(res) {
  const options = [];
  if (res.candidate) options.push(res.candidate);
  for (const c of res.alternatives || []) {
    if (options.length >= 5) break;
    if (!options.includes(c)) options.push(c);
  }
  return options;
}

function titleCase(s) {
  return String(s).replace(/[A-Za-z]+/g, (w) => w[0].toUpperCase() + w.slice(1).toLowerCase());
}

/** The fields review.py prints for each candidate, kept as structured data. */
export function describe(c) {
  const who = c.corporate
    || ((c.authors || []).slice(0, 3).map(titleCase).join(', ') || '?');
  const ident = c.doi || (c.pmid ? `PMID:${c.pmid}` : 'no identifier');
  return {
    title: c.title || '(untitled)',
    who,
    where: c.journal || c.publisher || '?',
    year: c.year || '?',
    locator: `v${c.volume || '?'}(${c.issue || '?'}):${c.first_page || '?'}`,
    ident,
    itemType: c.item_type,
    providers: [...(c.providers || [])].sort(),
  };
}

/** A Candidate built from a DOI the reader supplied, plus the parsed reference. */
export function candidateFromDoi(ref, doi) {
  return new Candidate({
    provider: 'manual',
    title: ref.title,
    doi,
    authors: [...ref.authors],
    corporate: ref.corporate,
    year: ref.year,
    journal: ref.journal,
    volume: ref.volume,
    issue: ref.issue,
    first_page: ref.first_page,
    last_page: ref.last_page,
    item_type: ref.is_book ? 'book' : 'journalArticle',
  });
}

/**
 * Apply one decision to a Resolution, in place.
 *
 * @param {'candidate'|'doi'|'from-text'|'skip'} kind
 * @param {object} payload  { candidate } | { doi } | {}
 */
export function applyDecision(res, ref, kind, payload = {}) {
  if (kind === 'skip') {
    res.status = 'REVIEW';
    res.tier = 'none';
    res.candidate = null;
    res.reviewed = 'skipped';
    return res;
  }
  if (kind === 'from-text') {
    res.candidate = itemFromText(ref);
    res.status = 'FROM_TEXT';
    res.tier = 'book-from-text';
    res.confidence = 0.80;
    res.reviewed = 'built from the reference text';
    return res;
  }
  if (kind === 'doi') {
    res.candidate = candidateFromDoi(ref, payload.doi);
    res.status = 'ACCEPTED';
    res.tier = 'manual';
    res.confidence = 1.0;
    res.reviewed = `accepted ${payload.doi}`;
    return res;
  }
  // a candidate the reader chose
  res.candidate = payload.candidate;
  res.status = 'ACCEPTED';
  res.tier = 'manual';
  res.confidence = 1.0;
  res.reviewed = `accepted ${payload.candidate.doi || payload.candidate.title.slice(0, 50)}`;
  return res;
}

/**
 * A DOI as the reader is likely to paste it: bare, prefixed, or as a URL.
 * Returns the normalised DOI, or null if it does not look like one at all.
 */
export function cleanDoi(input) {
  let d = String(input || '').trim();
  d = d.replace(/^https?:\/\/(dx\.)?doi\.org\//i, '');
  d = d.replace(/^doi:\s*/i, '');
  d = d.replace(/[.,;]+$/, '');
  return /^10\.\d{4,9}\/\S+$/.test(d) ? d : null;
}
